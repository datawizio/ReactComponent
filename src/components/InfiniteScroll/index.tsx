import * as React from "react";
import { useCallback, useState, useImperativeHandle } from "react";

import Spin from "antd/lib/spin";

import "./index.less";

const DefaultLoader = () => {
  return (
    <div className="infinite-scroll__default-loader">
      <Spin />
    </div>
  );
};

export interface InfiniteScrollProps {
  maxPage?: number;
  overflow?: string;
  showLoader?: boolean;
  triggerCoeff?: number;
  height: number | string;

  loader?: () => React.ReactNode;
  onBottom?: (page: number) => Promise<void> | void;
}

export interface InfiniteScrollRef {
  reset: () => void;
}

const InfiniteScroll = React.forwardRef<InfiniteScrollRef, InfiniteScrollProps>(
  (props, ref) => {
    const {
      height,
      loader,
      overflow,
      onBottom,
      children,
      showLoader,
      triggerCoeff,
      maxPage
    } = props;

    const [currentPage, setCurrentPage] = useState<number>(1);
    const [scrollIsHandling, setScrollIsHandling] = useState<boolean>(false);

    const scrollIsBottom = useCallback(
      target => {
        const triggerHeight = (target.scrollHeight * triggerCoeff) / 100;
        const triggerZone = target.clientHeight + triggerHeight;
        const scrollPos = target.scrollHeight - target.scrollTop;
        return triggerZone >= scrollPos;
      },
      [triggerCoeff]
    );

    function handleScroll(event) {
      if (scrollIsHandling) return;

      const handle = async handler => {
        setScrollIsHandling(true);
        await handler(currentPage);
        setScrollIsHandling(false);
      };

      if (onBottom && currentPage <= maxPage && scrollIsBottom(event.target))
        handle(onBottom).then(() => {
          setCurrentPage(currentPage + 1);
        });
    }

    useImperativeHandle(ref, () => ({
      reset() {
        setCurrentPage(1);
        setScrollIsHandling(false);
      }
    }));

    return (
      <div
        onScroll={handleScroll}
        className="infinite-scroll"
        style={{ height, overflow }}
      >
        {children}
        {showLoader && scrollIsHandling && loader()}
      </div>
    );
  }
);

InfiniteScroll.defaultProps = {
  triggerCoeff: 5,
  overflow: "auto",
  showLoader: false,
  maxPage: Infinity,
  loader: DefaultLoader
};

export default InfiniteScroll;
