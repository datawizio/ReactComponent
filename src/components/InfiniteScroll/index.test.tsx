import "jsdom-global/register";
import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import InfiniteScroll, { InfiniteScrollRef } from "./index";

const waitForComponentToPaint = async wrapper => {
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve));
    wrapper.update();
  });
};

const mockProps = {
  height: 550,
  maxPage: 3,
  onBottom: jest.fn(),
  triggerCoeff: 10,
  showLoader: true
};

const setUp = (props?) => mount(<InfiniteScroll {...props} />);

describe("InfiniteScroll component", () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);
  });

  it("Render InfiniteScroll correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("should scrolled and spin loader render and stoped scroll when handling", () => {
    const mEvent = {
      target: { scrollHeight: 100, scrollTop: 50, clientHeight: 50 }
    };
    waitForComponentToPaint(component);
    act(() => {
      component.find(".infinite-scroll").simulate("scroll", mEvent);
    });
    component.update();
    act(() => {
      component.find(".infinite-scroll").simulate("scroll", mEvent);
    });
    expect(component.find(".infinite-scroll__default-loader").length).toBe(1);
  });

  it("shoutld reset called", () => {
    const mEvent = {
      target: { scrollHeight: 100, scrollTop: 50, clientHeight: 50 }
    };
    const resetRef = React.createRef<InfiniteScrollRef>();
    const wrapper = mount(<InfiniteScroll {...mockProps} ref={resetRef} />);
    act(() => {
      wrapper.find(".infinite-scroll").simulate("scroll", mEvent);
    });
    wrapper.update();
    expect(wrapper.find(".infinite-scroll__default-loader").length).toBe(1);
    act(() => {
      resetRef.current.reset();
    });
    wrapper.update();
    expect(wrapper.find(".infinite-scroll__default-loader").length).toBe(0);
  });
});
