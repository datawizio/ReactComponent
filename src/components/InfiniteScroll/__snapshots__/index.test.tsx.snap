// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll component Render InfiniteScroll correctly 1`] = `
<ForwardRef
  height={550}
  loader={[Function]}
  maxPage={3}
  onBottom={[MockFunction]}
  overflow="auto"
  showLoader={true}
  triggerCoeff={true}
>
  <div
    className="infinite-scroll"
    onScroll={[Function]}
    style={
      Object {
        "height": 550,
        "overflow": "auto",
      }
    }
  />
</ForwardRef>
`;
